lightwood/encoders/time_series/transformer.py

```diff
@@ -0,0 +1,194 @@
+        losses = self._criterion(output, targets).sum(dim=1)
+        # The TBPTT will compute a slightly different loss, but it is not problematic
+        loss = torch.dot((1.0 / lengths.float()), losses) / len(losses)
+        return loss
+
+    def _prepare_raw_data(self, data):
+        # Convert to array and determine max length:
+        # priming_data is a list of len B with lists of length in [0, T]
+        # If the length of the series varies a lot, it would be worth it to order them and apply a few optimisations
+        # we do not do so now
+        out_data = [
+            torch.tensor(e, dtype=torch.float, device=self.device) for e in data
+        ]
+        lengths = torch.tensor(
+            [len(e) for e in data], dtype=torch.long, device=self.device
+        )
+        return out_data, lengths
+
+    def prepare_encoder(
+        self, priming_data, feedback_hoop_function=None, batch_size=256
+    ):
+        """
+        The usual, run this on the initial training data for the encoder
+        :param priming_data: a list of 1-dimensional time series [[list1], ..., [listT]]
+        :param feedback_hoop_function: [if you want to get feedback on the training process]
+        :param batch_size: int
+        :return:
+        """
+        if self._prepared:
+            raise RuntimeError(
+                'You can only call "prepare_encoder" once for a given encoder.'
+            )
+
+        # We append the eos symbol
+        self._append_eos(priming_data)
+        priming_data, lengths_data = self._prepare_raw_data(priming_data)
+        npoints = len(priming_data)
+
+        # We pad the sequence with the function from PyTorch
+        self._encoder.train()
+        for epoch in range(self._train_iters):
+            total_loss = 0.0
+            for start_batch in range(0, npoints, batch_size):
+                self._optimizer.zero_grad()
+
+                train_batch = self._get_batch(priming_data, start_batch, batch_size)
+                len_batch = self._get_batch(lengths_data, start_batch, batch_size)
+                train_batch = pad_sequence(train_batch)
+
+                # TBPTT
+                losses = []
+                for start_chunk in range(0, train_batch.size(0) - 1, self.bptt):
+                    data, targets, lengths_chunk = self._get_chunk(
+                        train_batch, len_batch, start_chunk, self.bptt
+                    )
+                    data = data.unsqueeze(-1)
+                    output = self._encoder(data, lengths_chunk)
+                    output = output.squeeze(-1)
+                    losses.append(
+                        self._masked_criterion(output, targets, lengths_chunk)
+                    )
+                loss = sum(losses)
+                loss.backward()
+                torch.nn.utils.clip_grad_norm_(self._encoder.parameters(), 0.5)
+                self._optimizer.step()
+
+                total_loss += loss.item() * len(train_batch)
```

Comment:
I'm not sure I get the `*len(train_batch)` operation here. `loss` holds the sum of the average loss in every chunk, right?

Reply:
At the moment we do stochastic gradient descent, and as such, we compute the batch loss as 1/B \sum_{i=1}^B \grad f(x_i, \theta). As such, we multiply by the batch size to recover the sum of the losses, to then compute the average loss per element of the whole dataset in the outside loop.